iceoryx_posh/include/iceoryx_posh/experimental/internal/popo/sample.inl

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IOX_EXPERIMENTAL_POSH_POPO_SAMPLE_INL
+#define IOX_EXPERIMENTAL_POSH_POPO_SAMPLE_INL
+
+namespace iox
+{
+namespace popo
+{
+
+// ============================== Sample<T> ========================= //
+template <typename T>
+Sample<T>::Sample(cxx::unique_ptr<T>&& samplePtr, PublisherInterface<T>& publisher)
+    : m_samplePtr(std::move(samplePtr))
+    , m_publisherRef(publisher)
+{
+    if(samplePtr)
+    {
+        m_hasOwnership = true;
+    }
+};
+
+template <typename T>
+Sample<T>::Sample(std::nullptr_t) noexcept {};
+
+template <typename T>
+Sample<T>& Sample<T>::operator=(Sample<T>&& rhs)
+{
+    if (this != &rhs)
+    {
+        m_samplePtr = std::move(rhs.m_samplePtr);
+        m_publisherRef = rhs.m_publisherRef;
+        m_hasOwnership = rhs.m_hasOwnership;
+        rhs.m_hasOwnership = false;
```

Comment:
I would add here `rhs.m_samplePtr = nullptr;` and maybe also turn the `m_publisherRef` into a pointer. The reason is to avoid undefined behavior.
When we set both to nullptr during move and you access a moved object you terminate right away with a segfault since you dereferenced a nullptr. I have no idea if this is possible with references. Otherwise the other pointer could somehow still access `m_publisherRef` since the value is still stored in there.

Reply:
Is it a problem for the publisher to still be accessible ? I don't think it's such a big issue... and it allows users to identify which publisher the sample originally came from (even though it is now empty and invalid).